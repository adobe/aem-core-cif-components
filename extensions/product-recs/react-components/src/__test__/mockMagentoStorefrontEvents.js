/*******************************************************************************
 *
 *    Copyright 2021 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

const methods = {
    context: {},
    publish: {}
};

const getMock = (prop, method) => {
    if (methods[prop][method]) {
        return methods[prop][method];
    }
    let mock = jest.fn();
    methods[prop][method] = mock;
    return mock;
};

const mockClear = () => {
    Object.keys(methods.context).forEach(k => methods.context[k].mockClear());
    Object.keys(methods.publish).forEach(k => methods.publish[k].mockClear());
};

const propHandler = {
    get: (target, prop) => {
        if (['context', 'publish'].includes(prop)) {
            const methodHandler = {
                get: (t, method) => {
                    return getMock(prop, method);
                }
            };
            return new Proxy({}, methodHandler);
        } else if (prop === 'mockClear') {
            return mockClear;
        }
    }
};

export default new Proxy({}, propHandler);
