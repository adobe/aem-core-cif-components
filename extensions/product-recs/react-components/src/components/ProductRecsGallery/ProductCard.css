/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ~ Copyright 2021 Adobe
 ~
 ~ Licensed under the Apache License, Version 2.0 (the "License");
 ~ you may not use this file except in compliance with the License.
 ~ You may obtain a copy of the License at
 ~
 ~     http://www.apache.org/licenses/LICENSE-2.0
 ~
 ~ Unless required by applicable law or agreed to in writing, software
 ~ distributed under the License is distributed on an "AS IS" BASIS,
 ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 ~ See the License for the specific language governing permissions and
 ~ limitations under the License.
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
.card {
    display: inline-block;
    border: 10px solid #fff;
    font-size: 20px;
    text-align: center;
    width: 240px;
}

.cardImage {
    height: 243px;
}

.addToCart {
    align-items: center;
    cursor: pointer;
    display: inline-flex;
    justify-content: center;
    line-height: 1;
    pointer-events: auto;
    text-align: center;
    white-space: nowrap;
    background: none;
    border: 1px solid rgb(var(--color));
    border-radius: 1.5rem;
    font-size: 0.75rem;
    font-weight: 600;
    height: 2rem;
    min-width: 7.5rem;
    padding: 0 1rem;
    transition-duration: 384ms;
    transition-property: background-color, color;
    transition-timing-function: var(--venia-anim-standard);
    --color: var(--venia-text);
    color: white;
    background-color: rgb(var(--color));
}

.addToCart:hover {
    color: white;
    background-color: rgb(var(--venia-teal));
}

.productImage {
    max-width: 240px;
    max-height: 240px;
}

.price {
    font-size: 0.875rem;
    line-height: 1rem;
    margin: 0.5rem 0;
    min-height: 1rem;
}

.price .regularPrice {
    text-decoration: line-through;
    font-size: 0.8rem;
}

.price .discountedPrice {
    font-weight: bold;
    color: $venia-teal;
    font-size: 1.2rem;
}
