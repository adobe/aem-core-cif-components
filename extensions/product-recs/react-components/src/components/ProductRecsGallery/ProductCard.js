/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ~ Copyright 2021 Adobe
 ~
 ~ Licensed under the Apache License, Version 2.0 (the "License");
 ~ you may not use this file except in compliance with the License.
 ~ You may obtain a copy of the License at
 ~
 ~     http://www.apache.org/licenses/LICENSE-2.0
 ~
 ~ Unless required by applicable law or agreed to in writing, software
 ~ distributed under the License is distributed on an "AS IS" BASIS,
 ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 ~ See the License for the specific language governing permissions and
 ~ limitations under the License.
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
import React from 'react';
import PropTypes from 'prop-types';

import { useIntl } from 'react-intl';
import { useStorefrontEvents, Price, Trigger, createProductPageUrl } from '@adobe/aem-core-cif-react-components';

import classes from './ProductCard.css';

const ProductCard = props => {
    const mse = useStorefrontEvents();
    const intl = useIntl();

    const {
        unit: { unitId },
        product: { sku, name, type, productId, currency, prices, smallImage }
    } = props;

    const addToCart = (unit, product) => {
        const { sku, type, productId } = product;
        const { unitId } = unit;

        const customEvent = new CustomEvent('aem.cif.add-to-cart', {
            detail: [{ sku, quantity: 1, virtual: type === 'virtual' }]
        });
        document.dispatchEvent(customEvent);

        mse && mse.publish.recsItemAddToCartClick(unitId, productId);
    };

    const renderPrice = (prices, currency) => {
        const { minimum, maximum } = prices;
        const isRange = Math.round(minimum.final * 100) !== Math.round(maximum.final * 100);

        if (isRange) {
            return (
                <>
                    {intl.formatMessage({ id: 'productrecs:price-from', defaultMessage: 'from' })}{' '}
                    <Price value={minimum.final} currencyCode={currency} />
                </>
            );
        }

        return <Price value={minimum.final} currencyCode={currency} />;
    };

    return (
        <div className={classes.card} key={sku}>
            <a
                href={createProductPageUrl(sku)}
                title={name}
                onClick={() => mse && mse.publish.recsItemClick(unitId, productId)}>
                <div className={classes.cardImage}>
                    <img className={classes.productImage} src={smallImage.url} alt={name} />
                </div>
                <div>{name}</div>
                <div className={classes.price}>{renderPrice(prices, currency)}</div>
            </a>
            {// Only display add to cart button for products that can be added to cart without further customization
            ['simple', 'virtual', 'downloadable'].includes(type) && (
                <Trigger action={() => addToCart(props.unit, props.product)}>
                    <span className={classes.addToCart}>
                        {intl.formatMessage({ id: 'productrecs:add-to-cart', defaultMessage: 'Add to cart' })}
                    </span>
                </Trigger>
            )}
        </div>
    );
};

ProductCard.propTypes = {
    unit: PropTypes.shape({
        unitId: PropTypes.string.isRequired
    }),
    product: PropTypes.shape({
        sku: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        type: PropTypes.string,
        productId: PropTypes.number,
        currency: PropTypes.string,
        prices: PropTypes.shape({
            minimum: PropTypes.shape({
                final: PropTypes.number
            }),
            maximum: PropTypes.shape({
                final: PropTypes.number
            })
        }),
        smallImage: PropTypes.shape({
            url: PropTypes.string
        })
    })
};

export default ProductCard;
