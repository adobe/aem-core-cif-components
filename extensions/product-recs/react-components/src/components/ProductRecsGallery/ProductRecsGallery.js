/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ~ Copyright 2021 Adobe
 ~
 ~ Licensed under the Apache License, Version 2.0 (the "License");
 ~ you may not use this file except in compliance with the License.
 ~ You may obtain a copy of the License at
 ~
 ~     http://www.apache.org/licenses/LICENSE-2.0
 ~
 ~ Unless required by applicable law or agreed to in writing, software
 ~ distributed under the License is distributed on an "AS IS" BASIS,
 ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 ~ See the License for the specific language governing permissions and
 ~ limitations under the License.
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
import React, { useRef } from 'react';
import PropTypes from 'prop-types';

import { useStorefrontEvents, LoadingIndicator } from '@adobe/aem-core-cif-react-components';

import { useRecommendations } from '../../hooks/useRecommendations';
import { useVisibilityObserver } from '../../hooks/useVisibilityObserver';

import classes from './ProductRecsGallery.css';
import ProductCard from './ProductCard';

const ProductRecsGallery = props => {
    const mse = useStorefrontEvents();
    const rendered = useRef(false);
    const { loading, units } = useRecommendations(props);
    const { observeElement } = useVisibilityObserver();

    let content = '';

    if (loading) {
        content = <LoadingIndicator />;
    }

    if (units && units.length > 0 && units[0].products.length > 0) {
        const unit = units[0];

        const isVisible = () => {
            mse && mse.publish.recsUnitView(unit.unitId);
        };

        content = (
            <>
                <h2 className={classes.title}>{unit.unitName || props.title}</h2>
                <div className={classes.container} ref={e => observeElement(e, isVisible)}>
                    {unit.products.map(product => (
                        <ProductCard unit={unit} product={product} key={product.sku} />
                    ))}
                </div>
            </>
        );

        if (!rendered.current) {
            mse && mse.publish.recsUnitRender(unit.unitId);
            rendered.current = true;
        }
    }

    return <div className={classes.root}>{content}</div>;
};

ProductRecsGallery.propTypes = {
    title: PropTypes.string,
    recommendationType: PropTypes.string,
    categoryExclusions: PropTypes.string,
    categoryInclusions: PropTypes.string,
    excludeMaxPrice: PropTypes.string,
    excludeMinPrice: PropTypes.string,
    includeMaxPrice: PropTypes.string,
    includeMinPrice: PropTypes.string,
    preconfigured: PropTypes.bool
};

export default ProductRecsGallery;
