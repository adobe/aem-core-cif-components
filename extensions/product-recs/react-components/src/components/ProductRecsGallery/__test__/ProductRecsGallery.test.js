/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ~ Copyright 2021 Adobe
 ~
 ~ Licensed under the Apache License, Version 2.0 (the "License");
 ~ you may not use this file except in compliance with the License.
 ~ You may obtain a copy of the License at
 ~
 ~     http://www.apache.org/licenses/LICENSE-2.0
 ~
 ~ Unless required by applicable law or agreed to in writing, software
 ~ distributed under the License is distributed on an "AS IS" BASIS,
 ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 ~ See the License for the specific language governing permissions and
 ~ limitations under the License.
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
import React from 'react';

import { render } from '@testing-library/react';

import ProductRecsGallery from '../ProductRecsGallery';
import mockMagentoStorefrontEvents from '../../../__test__/mockMagentoStorefrontEvents';
import ContextWrapper from '../../../__test__/context-wrapper';

const mockUseRecommendationsValue = jest.fn();
jest.mock('../../../hooks/useRecommendations', () => ({
    useRecommendations: () => mockUseRecommendationsValue()
}));
jest.mock('../../../hooks/useVisibilityObserver', () => ({
    useVisibilityObserver: () => ({ observeElement: jest.fn() })
}));

describe('ProductRecsGallery', () => {
    let mse;

    beforeAll(() => {
        mse = window.magentoStorefrontEvents = mockMagentoStorefrontEvents;
    });

    beforeEach(() => {
        window.magentoStorefrontEvents.mockClear();
    });

    beforeEach(() => {
        mockUseRecommendationsValue.mockClear();
    });

    it('renders a loading state', () => {
        mockUseRecommendationsValue.mockReturnValue({ loading: true, units: null });

        const { asFragment } = render(
            <ProductRecsGallery title="My Product Recommendations" recommendationType="most-viewed" />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('renders a list of products', () => {
        mockUseRecommendationsValue.mockReturnValue({
            loading: false,
            units: [
                {
                    unitId: 'my-unit-id',
                    products: [
                        {
                            sku: 'sku-a',
                            name: 'My Product A',
                            type: 'simple',
                            productId: 1,
                            currency: 'CHF',
                            prices: {
                                minimum: {
                                    final: 342.23
                                },
                                maximum: {
                                    final: 342.23
                                }
                            },
                            smallImage: {
                                url: 'http://localhost/product-a.jpg'
                            }
                        },
                        {
                            sku: 'sku-b',
                            name: 'My Product B',
                            type: 'simple',
                            productId: 2,
                            currency: 'CHF',
                            prices: {
                                minimum: {
                                    final: 342.23
                                },
                                maximum: {
                                    final: 2231.42
                                }
                            },
                            smallImage: {
                                url: 'http://localhost/product-b.png'
                            }
                        }
                    ]
                }
            ]
        });

        const { asFragment } = render(
            <ProductRecsGallery title="My Product Recommendations" recommendationType="most-viewed" />,
            { wrapper: ContextWrapper }
        );

        expect(asFragment()).toMatchSnapshot();
        expect(mse.publish.recsUnitRender).toHaveBeenCalledWith('my-unit-id');
    });
});
