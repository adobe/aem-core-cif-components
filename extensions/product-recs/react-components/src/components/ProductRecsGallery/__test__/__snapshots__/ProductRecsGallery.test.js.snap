// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductRecsGallery renders a list of products 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <h2
      class="title"
    >
      My Product Recommendations
    </h2>
    <div
      class="container"
    >
      <div
        class="card"
      >
        <a
          title="My Product A"
        >
          <div
            class="cardImage"
          >
            <img
              alt="My Product A"
              class="productImage"
              src="http://localhost/product-a.jpg"
            />
          </div>
          <div>
            My Product A
          </div>
          <div
            class="price"
          >
            <span
              class=""
            >
              CHF 342.23
            </span>
          </div>
        </a>
        <button
          class="cmp-Trigger__trigger__root cmp-components__clickable__root"
          type="button"
        >
          <span
            class="addToCart"
          >
            Add to cart
          </span>
        </button>
      </div>
      <div
        class="card"
      >
        <a
          title="My Product B"
        >
          <div
            class="cardImage"
          >
            <img
              alt="My Product B"
              class="productImage"
              src="http://localhost/product-b.png"
            />
          </div>
          <div>
            My Product B
          </div>
          <div
            class="price"
          >
            from 
            <span
              class=""
            >
              CHF 342.23
            </span>
          </div>
        </a>
        <button
          class="cmp-Trigger__trigger__root cmp-components__clickable__root"
          type="button"
        >
          <span
            class="addToCart"
          >
            Add to cart
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProductRecsGallery renders a loading state 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="cmp-LoadingIndicator__indicator__root"
    >
      <span
        class="cmp-LoadingIndicator__indicator__indicator loader-img"
      />
      <span
        class="cmp-LoadingIndicator__indicator__message"
      />
    </div>
  </div>
</DocumentFragment>
`;
